File: Typescript/InterpreterPattern.ts
// Abstract Expression class
interface Expression {
  interpret(): number;
}

// Terminal Expression for numbers
class NumberExpression implements Expression {
  constructor(private value: number) {}

  interpret(): number {
    return this.value;
  }
}

// Non-terminal Expression for addition operation
class AdditionExpression implements Expression {
  constructor(
    private expression1: Expression,
    private expression2: Expression
  ) {}

  interpret(): number {
    return this.expression1.interpret() + this.expression2.interpret();
  }
}

// Non-terminal Expression for subtraction operation
class SubtractionExpression implements Expression {
  constructor(
    private expression1: Expression,
    private expression2: Expression
  ) {}

  interpret(): number {
    return this.expression1.interpret() - this.expression2.interpret();
  }
}

// Example usage
// Construct expressions: 10 + (20 - 5)
const expression: Expression = new AdditionExpression(
  new NumberExpression(10),
  new SubtractionExpression(new NumberExpression(20), new NumberExpression(5))
);

// Interpret and evaluate the expression
const result: number = expression.interpret();
console.log("Result:", result);

File: Python/InterpreterPattern.py
# Abstract Expression class
class Expression:
    def interpret(self):
        pass

# Terminal Expression for numbers
class NumberExpression(Expression):
    def __init__(self, value):
        self.value = value

    def interpret(self):
        return self.value

# Non-terminal Expression for addition operation
class AdditionExpression(Expression):
    def __init__(self, expression1, expression2):
        self.expression1 = expression1
        self.expression2 = expression2

    def interpret(self):
        return self.expression1.interpret() + self.expression2.interpret()

# Non-terminal Expression for subtraction operation
class SubtractionExpression(Expression):
    def __init__(self, expression1, expression2):
        self.expression1 = expression1
        self.expression2 = expression2

    def interpret(self):
        return self.expression1.interpret() - self.expression2.interpret()

# Example usage
if __name__ == "__main__":
    # Construct expressions: 10 + (20 - 5)
    expression = AdditionExpression(NumberExpression(10), SubtractionExpression(NumberExpression(20), NumberExpression(5)))

    # Interpret and evaluate the expression
    result = expression.interpret()
    print("Result:", result)
